src/config/proxy.rs

```diff
@@ -0,0 +1,169 @@
+use http::uri::{InvalidUri, Uri};
+use hyper_proxy::{Custom, Intercept, Proxy, ProxyConnector};
+use std::collections::HashSet;
+
+fn from_env(key: &str) -> Option<String> {
+    std::env::var(key.to_string())
+        .ok()
+        .or_else(|| std::env::var(key.to_lowercase()).ok())
+}
+
+struct NoProxyCache(HashSet<Uri>);
+
+impl From<&HashSet<String>> for NoProxyCache {
+    fn from(hashset: &HashSet<String>) -> Self {
+        Self(
+            hashset
+                .iter()
+                .filter_map(|uri| uri.parse::<Uri>().ok())
+                .collect(),
+        )
+    }
+}
+
+impl NoProxyCache {
+    fn matches(&self, scheme: Option<&str>, host: Option<&str>, port: Option<u16>) -> bool {
+        for uri in self.0.iter() {
+            if uri.scheme_str() == scheme && uri.host() == host && uri.port_u16() == port {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn intercept(self, expected_scheme: &'static str) -> Intercept {
+        Intercept::Custom(Custom::from(
+            move |scheme: Option<&str>, host: Option<&str>, port: Option<u16>| {
+                if scheme != Some(expected_scheme) {
+                    return false;
+                }
+                // only intercapt those that don't match
+                !self.matches(scheme, host, port)
+            },
+        ))
+    }
+}
+
+#[derive(serde::Deserialize, serde::Serialize, Clone, Default, Debug, PartialEq)]
+pub struct ProxyConfig {
+    pub http: Option<String>,
+    pub https: Option<String>,
+    pub no_proxy: HashSet<String>,
+}
+
+impl ProxyConfig {
+    pub fn from_env() -> Self {
+        Self {
+            http: from_env("HTTP_PROXY"),
+            https: from_env("HTTP_PROXYS"),
```

Comment:
```suggestion
            https: from_env("HTTPS_PROXY"),
```

Reply:
Yeah, that is a definite typo which does seem to indicate a lack of test coverage here. Could we fix the typo and add tests for it?